tests/util/bson.cpp

```diff
@@ -0,0 +1,411 @@
+#include "catch2/catch.hpp"
+#include <stdio.h>
+#include <json.hpp>
+#include "util/test_utils.hpp"
+#include "util/test_file.hpp"
+#include "util/bson.hpp"
+#include "util/extended_json.hpp"
+#include <variant>
+#include <any>
+
+using namespace nlohmann;
+using namespace realm;
+using namespace bson;
+
+static inline std::string remove_whitespace(const char* c) {
+    std::string str(c);
+    str.erase(std::remove_if(str.begin(), str.end(), isspace), str.end());
+    return str;
+}
+
+/**
+ ======== BSON CORPUS ========
+ */
+template <typename T>
+using corpus_check_t = std::function<bool(T)>;
+
+template <typename T>
+struct corpus_entry_t {
+    const char* description;
+    const char* canonical_extjson;
+    corpus_check_t<T> check;
+    bool lossy;
+};
+
+template <typename T>
+using corpus_t = std::vector<corpus_entry_t<T>>;
+
+template <typename T>
+static inline void run_corpus(const char* test_key, const corpus_t<T>& corpus) {
```

Comment:
This function mostly just obfuscates the tests and makes them harder to debug. Eliminating the loop and calling the function once per SECTION would be basically the same amount of code, but simpler.